pkg/docker/concurrent.go

```diff
@@ -0,0 +1,137 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+type ConcurrentImageProcessor struct {
+	maxRoutines int
+}
+
+func NewConcurrentImageProcessor(maxRoutines int) *ConcurrentImageProcessor {
+	return &ConcurrentImageProcessor{maxRoutines: maxRoutines}
+}
+
+type ImageProcessor func(ctx context.Context, image string) error
+
+func (c *ConcurrentImageProcessor) Process(ctx context.Context, images []string, process ImageProcessor) error {
+	contextWithCancel, abortRemainingJobs := context.WithCancel(ctx)
+	workers := min(len(images), c.maxRoutines)
+	workers = max(workers, 1)
+	jobsChan := make(chan job)
+	wg := &sync.WaitGroup{}
+	errors := make(chan error)
+	doneChan := make(chan struct{})
+
+	for i := 0; i < workers; i++ {
+		w := &worker{
+			jobs:        jobsChan,
+			process:     process,
+			waitGroup:   wg,
+			errorReturn: errors,
+		}
+		go w.start()
+	}
+
+	f := &feeder{
+		jobs:            jobsChan,
+		images:          images,
+		workersWaiGroup: wg,
+		done:            doneChan,
+	}
+	go f.feed(contextWithCancel)
+
+	var firstError error
+
+loop:
+	for {
+		select {
+		case <-doneChan:
+			break loop
+		case err := <-errors:
+			if firstError == nil {
```

Comment:
curious why this is needed? can't we run `abortRemainingJobs()` whenever there's an err?

Reply:
Great question
There is a race condition: 
* One worker fails
* The error is sent through the errors channel
* The main routine reads it and cancels the context
* The feeder stops feeding tasks, closes the channel and waits for all workers to finish the current tasks.
* Since the tasks also receive the context, depending on the implementation, they could be stopped when the context is cancelled. This will return an error through the errors channel.
* If this error is returned before the `doneChan` is closed, the main routine will receive it as well.
* If we don't preserve the first error, it will be overwritten by this second error and we lose the original cause of the problem

So answering you question, yes, we can run `abortRemainingJobs()` whenever. But we don't want to overwrite the first error.

Reply:
i see, make sense